### ISSUE TITLE
Breadcrumbs is not easily customizable

### ISSUE THREAD

**Issue by vlaurier** [sonata-project/SonataAdminBundle#4640]
#### Sonata packages
```
"sonata-project/admin-bundle":  "3.19"
```
#### Symfony packages
```
"symfony/symfony": "2.8.21"
```
#### PHP version
```
PHP 5.6
```
## Subject

Issue : The breadcrumbs is not easily customizable.

Explanation: 

The class [BreadcrumbsBuilder](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/Admin/BreadcrumbsBuilder.php) is Final, and the service is directly retrieved from container in sonata [CRUDController](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/Controller/CRUDController.php)

Then the variable "breadcrumbs_builder" is used inside the [standard layout](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/Resources/views/standard_layout.html.twig)

As a consequence, the implementation is closed to extension.
How is it possible to add custom logic for breadcrumbs ?

I thought it was relevant to point this architecture issue if someone wants to propose a more open implementation.

**Comment by greg0ire** [comment]
I did that change, on purpose. Related :
- https://ocramius.github.io/blog/when-to-declare-classes-final/
- http://fabien.potencier.org/pragmatism-over-theory-protected-vs-private.html

TL;DR : tell us what you are trying to achieve, what you would like to customize exactly, and maybe we can create a new but smaller extension point that can benefit everyone rather than having people just override everything and keep the improvements they do in their codebase.

**Comment by vlaurier** [comment]
Thanks for your reply. I try to give some more explanations for my need.
I wanted for some admin class to have a custom breadcrumbs depending on the subject they manage.

For example, one of them should show an identifier at the end of the breadcrumbs, and all category ancestors, which make sense in the business need.
Another one should show a title, which is a concatenation of some subject's data, and add a text in the breadcrumbs.

I ended up by registering my own breadcrumbsBuilder with the sonata service id.
And this builder calls some custom builders for specific cases. At the end it calls the sonata one by default.

**Comment by greg0ire** [comment]
> For example, one of them should show an identifier at the end of the breadcrumbs, and all category ancestors, which make sense in the business need.

Not completely sure what this means, maybe an example would help?



>  Another one should show a title, which is a concatenation of some subject's data, and add a text in the breadcrumbs.

I think that one can be done by overriding `AbstractAdmin::toString()`, but maybe you didn't want to have that change everywhere?

Maybe you could have achieved the first thing with that method too, but didn't, for the same reason? In that case, it would be great to contribute something that would allow to customize the last item by providing a specialized service which would have that job.

**Comment by vlaurier** [comment]
I give an example for the first case:

Home > GrandParentCategory > ParentCategory > Category > MyIdentifier

where categories can be retrieved from the subject of some specific admin classes.

For the second case:

Home > the same beginning breadcrumb as Sonata one > A custom text here > A custom title here.

where a custom title also depends of the subject behind the admin class.

I hope it is clearer with these examples.

**Comment by yamitenshi** [comment]
I second this issue. I'd like to add a dynamically generated breadcrumb item to some admins, which should be relatively easy, but it's unnecessarily difficult.

I can understand closing the BreadcrumbsBuilder class for extension - I'm actually all in favour of making everything final unless extension is a requirement.

What mostly makes this harder, however, is that the breadcrumbs builder is being fetched _by a hard-coded service ID_ in the render() method of the CRUDController. This has not only closed the breadcrumbs builder for extension, but has effectively cut off any possibility for making use of the dependency injection container for its intended purpose.

My only options at this point are either overriding the CRUDController and _completely_ overwriting the `render()` method (which I really shouldn't have to do just for a single breadcrumb in a single admin), or completely overwriting Sonata's service definition (making my own breadcrumb builder apply to _every_ admin class, including those it was never intended for).

In my eyes, a possible solution would be fetching the breadcrumb builder in a separate CRUDController method (say `getBreadcrumbsBuilder()` as it used to be present in the Admin class) so we can override just that functionality where we need it.

**Comment by greg0ire** [comment]
> This has not only closed the breadcrumbs builder for extension, but has effectively cut off any possibility for making use of the dependency injection container for its intended purpose.

Wrong. See https://symfony.com/doc/3.2/service_container/service_decoration.html

**Comment by yamitenshi** [comment]
I did not know about that, thank!

That does allow me to easily overwrite the breadcrumbs builder, but it would still mean applying a breadcrumbs builder I need for a very specific purpose in a very specific place over _all_ admin classes.

It works, I guess, but it does seem like a suboptimal solution.

**Comment by greg0ire** [comment]
I have no solution for this one, sadly... I think the way to go for this would be to have sonata admin create one alias per admin, and to allow the user to change that alias to what they want for any admin they want. What do you think?

**Comment by yamitenshi** [comment]
That seems like a very good solution to me. Though I do admit, it's a harder problem to solve than I initially would have thought (as is so often the case 😄).

I'm not sure what the performance impact of creating a lot of aliases would be though (I'd imagine it not to be that big, especially in production, with proper caching). And either way, I can't really think of another solution that doesn't at some point involve stuffing functionality where it doesn't belong...

**Comment by greg0ire** [comment]
> I'm not sure what the performance impact of creating a lot of aliases would be though (I'd imagine it not to be that big, especially in production, with proper caching).

I think the impact would be really tiny. To implement the solution, I think all is needed is a new compiler pass that would be able to gather all admin services definitions, and create one alias for each of them. The alias could be changed to point to whatever you want by just adding something in the service definition tag.

cc @sagikazarmark with which I remember discussing similar things here: https://github.com/sonata-project/SonataAdminBundle/issues/3947

**Comment by greg0ire** [comment]
An alternative could be to tag your new service with the admin code, to tell the compiler pass that it needs to alias it to that service for that admin code.

**Comment by greg0ire** [comment]
cc @fbourigault who worked on extracting another component

**Comment by fbourigault** [comment]
The aliases solution would work. There would be no overhead as those are resolved at container build time. As `AbstractAdmin::setBreadcrumbBuilder` is deprecated since 3.2, I would introduce some registry or service locator to get the right breadcrumb builder for an admin and configure the service to use on the `sonata.admin` tag.
